src/lr_scheduler/mod.rs

```diff
@@ -0,0 +1,78 @@
+use crate::optimizer::Optimizer;
+use crate::value::Value;
+
+mod constant_lr;
+mod cosine_annealing_lr;
+mod exponential_lr;
+mod lambda_lr;
+mod linear_lr;
+mod multiplicative_lr;
+mod multistep_lr;
+mod polynomial_lr;
+mod step_lr;
+
+pub use self::constant_lr::ConstantLR;
+pub use self::cosine_annealing_lr::CosineAnnealingLR;
+pub use self::exponential_lr::ExponentialLR;
+pub use self::lambda_lr::LambdaLR;
+pub use self::linear_lr::LinearLR;
+pub use self::multiplicative_lr::MultiplicativeLR;
+pub use self::multistep_lr::MultiStepLR;
+pub use self::polynomial_lr::PolynomialLR;
+pub use self::step_lr::StepLR;
+
+pub trait Schedule {
+    const HAS_CLOSED_FORM: bool = true;
+
+    fn get_lr(&self, base_lr: f64, last_epoch: usize) -> f64;
+    fn get_closed_form_lr(&self, _base_lr: f64, _last_epoch: usize) -> f64 {
+        0.0
+    }
+}
+
+pub struct LRScheduler<S: Schedule> {
+    lr: Value,
+    schedule: S,
+    base_lr: f64,
+    last_epoch: usize,
+}
+
+impl<S: Schedule> LRScheduler<S> {
+    pub fn new<O: Optimizer>(optimizer: &O, schedule: S) -> Self {
+        let lr = optimizer.lr();
+        let base_lr = lr.value();
+        let last_epoch = 0;
```

Comment:
from what I have seen, Pytorch has a default `last_epoch` value of -1, is there a reason behind using 0 here?

Reply:
I made this choice early on and back then it felt like it made more sense to have `last_epoch` be a `usize` since -1 is treated as a sort of sentinel value in PyTorch. I compensated by having `step` change `lr` first and _then_ increment `last_epoch` (as opposed to PyTorch where it’s incremented first).

In retrospect, I didn’t look too far into why it’s -1 by default, and it’s likely that there’s some use case for it that I missed.

Reply:
Since `LRScheduler` doesn't have `last_epoch` as public it could be set to -1 but if this is a workaround thé keeping it as `usize` would be really nice

Reply:
I'm gonna review the PyTorch source code again and see if there's a good reason for it being -1 by default.

Reply:
Couldn't find anything concrete, though I think we look into setting the default as -1 in a future PR that introduces more complex schedulers.